core/internal/messagelog/messagelog_test.go

```diff
@@ -80,6 +80,86 @@ func TestStream(t *testing.T) {
 	assert.False(t, more, "Channel not closed")
 }
 
+func TestMessages(t *testing.T) {
+	const nrMessages = 5
+
+	log := New()
+	msgs := makeManyMsgs(nrMessages)
+
+	for _, msg := range msgs {
+		log.Append(msg)
+	}
+	assert.Equalf(t, msgs, log.Messages(), "Unexpected messages")
+}
+
+func TestReset(t *testing.T) {
+	const nrMessages = 23
+
+	log := New()
+	ch := log.Stream(nil)
+
+	msgs := makeManyMsgs(nrMessages)
+	log.Reset(msgs)
+
+	for i, m := range msgs {
+		assert.Equalf(t, m, <-ch, "Unexpected message %d", i)
+	}
+
+	msgs2 := makeManyMsgs(nrMessages)
+	log.Reset(msgs2)
+
+	ch2 := log.Stream(nil)
+	for i, m := range msgs2 {
+		assert.Equalf(t, m, <-ch, "Unexpected message %d", i)
+		assert.Equalf(t, m, <-ch2, "Unexpected message %d", i)
+	}
+	assert.Equalf(t, msgs2, log.Messages(), "Unexpected messages")
+}
+
+func TestResetConcurrent(t *testing.T) {
+	const nrStreams = 3
+	const nrMessages = 23
+
+	log := New()
+
+	msgs := makeManyMsgs(nrMessages)
+	msgs2 := makeManyMsgs(nrMessages)
+	log.Reset(msgs)
+
+	wg := new(sync.WaitGroup)
+	wg.Add(nrStreams)
+	for id := 0; id < nrStreams; id++ {
+		ch := log.Stream(nil)
+
+		go func(streamID int) {
+			defer wg.Done()
+
+			var i, j int
+			for j < len(msgs2) {
+				ok := assert.Conditionf(t, func() bool {
+					m := <-ch
+					switch {
+					case i < len(msgs) && m == msgs[i]:
+						i++
+						return assert.Zero(t, j)
+					case m == msgs2[j]:
+						j++
+						return true
+					default:
+						return false
+					}
+				}, "Unexpected message from stream %d", streamID)
+				if !ok {
+					break
+				}
+			}
+		}(id)
+	}
+
+	log.Reset(msgs2)
```

Comment:
It seems likely that the log is reset with msgs2 before the goroutines receive messages from msgs. So is it better to put a short delay before this reset?

Reply:
If we add a delay then the goroutines would likely receive all messages in `msgs` before the log is reset with `msgs2`.

Reply:
OK, so maybe the delay is not a good enough idea, I have no better option so I'm fine with current code.